source/extensions/filters/network/sni_dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -0,0 +1,76 @@
+#include "extensions/filters/network/sni_dynamic_forward_proxy/proxy_filter.h"
+
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/upstream/thread_local_cluster.h"
+
+#include "common/common/assert.h"
+#include "common/tcp_proxy/tcp_proxy.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SniDynamicForwardProxy {
+
+ProxyFilterConfig::ProxyFilterConfig(
+    const FilterConfig& proto_config,
+    Extensions::Common::DynamicForwardProxy::DnsCacheManagerFactory& cache_manager_factory,
+    Upstream::ClusterManager& cluster_manager)
+    : port_(static_cast<uint16_t>(proto_config.port_value())),
+      dns_cache_manager_(cache_manager_factory.get()),
+      dns_cache_(dns_cache_manager_->getCache(proto_config.dns_cache_config())),
+      cluster_manager_(cluster_manager) {}
+
+ProxyFilter::ProxyFilter(ProxyFilterConfigSharedPtr config) : config_(std::move(config)) {}
+
+using LoadDnsCacheEntryStatus = Common::DynamicForwardProxy::DnsCache::LoadDnsCacheEntryStatus;
+
+Network::FilterStatus ProxyFilter::onNewConnection() {
+  absl::string_view sni = read_callbacks_->connection().requestedServerName();
+  ENVOY_CONN_LOG(trace, "sni_dynamic_forward_proxy: new connection with server name '{}'",
+                 read_callbacks_->connection(), sni);
+
+  if (sni.empty()) {
+    return Network::FilterStatus::Continue;
+  }
+
+  // TODO(lizan): implement circuit breaker in SNI dynamic forward proxy like it is in HTTP:
```

Comment:
Is there an reason not to do this now since it's pretty easy? Or can this be done as an immediate follow up? This is going to get used and this is a pretty big potential issue I think?

Reply:
Unlike HTTP in network filters we don't have immediate access to cluster info through filter callbacks. There are plumbings need to happen there.

Reply:
OK sounds good.